app/hooks/useKeyDown.js

```diff
@@ -0,0 +1,69 @@
+// @flow
+import * as React from "react";
+import isTextInput from "utils/isTextInput";
+
+export type KeyFilter = ((event: KeyboardEvent) => boolean) | string;
+
+// Registered keyboard event callbacks
+let callbacks = [];
+
+// Track if IME input suggestions are open so we can ignore keydown shortcuts
+// in this case, they should never be triggered from mobile keyboards.
+let imeOpen = false;
+
+// Based on implementation in react-use
+// https://github.com/streamich/react-use/blob/master/src/useKey.ts#L15-L22
+const createKeyPredicate = (keyFilter: KeyFilter) =>
+  typeof keyFilter === "function"
+    ? keyFilter
+    : typeof keyFilter === "string"
+    ? (event: KeyboardEvent) => event.key === keyFilter
+    : keyFilter
+    ? (_event) => true
+    : (_event) => false;
+
+export default function useKeyDown(
+  key: KeyFilter,
+  fn: (event: KeyboardEvent) => void,
+  deps?: $ReadOnlyArray<mixed> = [key]
+): void {
+  const predicate = createKeyPredicate(key);
+
+  const handler = React.useCallback((event: KeyboardEvent) => {
+    console.log(event);
+    if (predicate(event)) {
+      fn(event);
+    }
+  }, deps);
+
+  React.useEffect(() => {
+    callbacks.push(handler);
+
+    return () => {
+      callbacks = callbacks.filter((cb) => cb !== handler);
+    };
+  });
+}
```

Comment:
I have a doubt here. Why do we have deps props? My best guess is it's because we want to memorize the handler but it doesn't seem to work because we are anyway pushing `handler` to `callbacks` every render.  

It feels kind of hard to read. If `fn` is memorized then why do we need `deps`? if not, then fn should always be one of the deps. 

We can move all the handler logic inside `useEffect`. What do you think of this?

```js
  React.useEffect(() => {
    const predicate = createKeyPredicate(key);
    const handler = (event: KeyboardEvent) => {
      if (predicate(event)) fn(event);
    };
    callbacks.push(handler);

    return () => {
      callbacks = callbacks.filter((cb) => cb !== handler);
    };
  });
```

Reply:
I haven't actually needed `deps` yet, so could potentially just remove that complexity until it's needed for something…